src/main/java/actionlogger/Dump.java

```diff
@@ -0,0 +1,154 @@
+            return;
+        }
+
+        var sceneData = new SceneData(scene.getBaseX(), scene.getBaseY(), scene.isInstance());
+        List<GroundItemData> groundItemData = new ArrayList<>();
+        List<ObjectData> objectData = new ArrayList<>();
+        List<NPCData> npcData = new ArrayList<>();
+        var tiles = scene.getTiles();
+        for (var z = 0; z < tiles.length; z++) {
+            for (var x = 0; x < tiles[z].length; x++) {
+                for (var y = 0; y < tiles[z][x].length; y++) {
+                    var tile = tiles[z][x][y];
+
+                    if (tile == null) {
+                        continue;
+                    }
+
+                    var groundItems = tile.getGroundItems();
+                    if (groundItems != null) {
+                        for (var groundItem : groundItems) {
+                            var itemComposition = itemManager.getItemComposition(groundItem.getId());
+                            groundItemData.add(new GroundItemData(x, y, z, itemComposition.getId(), itemComposition.getName(), groundItem.getQuantity()));
+                        }
+                    }
+
+                    var decorativeObjects = tile.getDecorativeObject();
+                    if (decorativeObjects != null) {
+                        objectData.add(new ObjectData(x, y, z, decorativeObjects.getId(), "DECORATIVE"));
+                    }
+
+                    var groundObject = tile.getGroundObject();
+                    if (groundObject != null) {
+                        objectData.add(new ObjectData(x, y, z, groundObject.getId(), "GROUND"));
+                    }
+
+                    var wallObject = tile.getWallObject();
+                    if (wallObject != null) {
+                        objectData.add(new ObjectData(x, y, z, wallObject.getId(), "WALL"));
+                    }
+
+                    var gameObjects = tile.getGameObjects();
+                    if (gameObjects != null) {
+                        for (var gameObject : gameObjects) {
+                            if (gameObject != null) {
+                                objectData.add(new ObjectData(x, y, z, gameObject.getId(), "GAME"));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        for (var npc : worldView.npcs()) {
+            var localLocation = npc.getLocalLocation();
+            npcData.add(new NPCData(localLocation.getSceneX(), localLocation.getSceneY(), worldView.getPlane(), npc.getId(), npc.getName()));
+        }
+
+        var dumpData = new DumpData(
+            sceneData,
+            worldView.getPlane(),
+            dumps.contains("grounditems") ? groundItemData : null,
+            dumps.contains("objects") ? objectData : null,
+            dumps.contains("npcs") ? npcData : null
+        );
+
+        writer.write("DUMP", dumpData);
+    }
+
+    @Data
```

Comment:
lombok tip: if you use `@Value`, you can omit `private final`

i only use `@Data` if the class contains a mutable field